src/main/java/fi/okm/jod/yksilo/repository/KoulutusmahdollisuusRepository.java

```diff
@@ -57,34 +58,60 @@ default List<MahdollisuusDto> findMahdollisuusIds(Kieli lang, Sort.Direction dir
           case SV -> "sv-x-icu";
           case EN -> "en-x-icu";
         };
+    // Cache
     // collate() is HQL extension
     final List<Tuple> mahdollisuusIdsImpl =
         findMahdollisuusIdsImpl(
             lang, JpaSort.unsafe(direction, "collate(m.otsikko as `" + collation + "`)"));
-    return mahdollisuusIdsImpl.stream()
+    Map<UUID, List<Tuple>> grouped =
+        mahdollisuusIdsImpl.stream().collect(Collectors.groupingBy(t -> t.get("id", UUID.class)));
+
+    return grouped.entrySet().stream()
         .map(
-            t ->
-                new MahdollisuusDto(
-                    t.get("id", UUID.class),
-                    t.get("tyyppi", String.class),
-                    t.get("ammattiryhma", String.class),
-                    t.get("aineisto", String.class),
-                    t.get("koulutusTyyppi", String.class)
-                    // Convert to enum inside DTO constructor if needed
-                    ))
-        .collect(Collectors.toList());
+            entry -> {
+              UUID id = entry.getKey();
+              List<Tuple> tuples = entry.getValue();
+              String tyyppi = tuples.getFirst().get("tyyppi", String.class);
+              String ammattiryhma = tuples.getFirst().get("ammattiryhma", String.class);
+              String aineisto = tuples.getFirst().get("aineisto", String.class);
+              String koulutusTyyppi = tuples.getFirst().get("koulutusTyyppi", String.class);
+              List<String> maakunnat =
```

Comment:
Our coding style convention prefers var where explicit type is not needed. 
(If using Idea, File | Settings | Editor | Code Style | Java -> Code Generation -> Use 'var' for local variable declaration)